tests/test_e2e/test_cloud_spec.py

```diff
@@ -0,0 +1,48 @@
+import ops
+import pytest
+
+import scenario
+from scenario.state import CloudSpec
+
+
+@pytest.fixture(scope="function")
+def mycharm():
+    class MyCharm(ops.CharmBase):
+        def __init__(self, framework: ops.Framework):
+            super().__init__(framework)
+            for evt in self.on.events().values():
+                self.framework.observe(evt, self._on_event)
+
+        def _on_event(self, event):
+            pass
+
+    return MyCharm
+
+
+def test_get_cloud_spec(mycharm):
+    cloud_spec = CloudSpec.from_dict(
+        {
+            "name": "localhost",
+            "type": "lxd",
+            "endpoint": "https://127.0.0.1:8443",
+            "credential": {
+                "auth-type": "certificate",
+                "attrs": {
+                    "client-cert": "foo",
+                    "client-key": "bar",
+                    "server-cert": "baz",
+                },
+            },
+        }
+    )
```

Comment:
And how would we feel about:
```
   cloud_spec = CloudSpec(
            name="localhost",
            type="lxd",
            endpoint="https://127.0.0.1:8443",
            auth_type: str | None = "certificate",
            client_cert: str | None = "foo",
            client_key: str | None = "bar",
            server-cert: str | None = "baz",
)
```

Reply:
I assume the users of OF and scenario are the same group, not that developers use OF only and testers use scenario only, right? If that is the case, I'd strongly argue that constructing a CloudSpec for testing purposes from a dict is easier from the user's standpoint for two reasons:

1. (As mentioned in another comment,) the user has already got a dict/json for UT (probably got by running the credential_get hook tool, or simply made-up). Re-use the same for both UT and scenario is simpler.
2. When creating UT, the user does not have to figure out the implementation of CloudSpec, what attributes are mandatory and what not. They simply copy a dict from credential_get hook tool output and use that. It's less cognitive load for the users. I mean, if they need to know what attributes there are in CloudSpec and create by instantiating anyway, they might as well do the same in UT, why have from_dict in UT in the first place? It doesn't save you much trouble because you will need to figure it out later in scenario anyway.